### ISSUE TITLE
[Bug][CPU][Spec Decode]: CPU expand_kernel shim discards its output — non-greedy spec decode consumes uninitialized memory as temperature/top_k/top_p

### ISSUE THREAD

**Issue by mosafariuk** [vllm-project/vllm#55005]
### Summary

The CPU monkey-patch for `expand_kernel` (`vllm/utils/cpu_triton_utils.py:408-421`) passes `_ensure_int64(output)` to the int64-only C++ kernel (`csrc/cpu/spec_decode_utils.cpp:405-428`). When the caller's buffer is not int64, `_ensure_int64` returns a **new tensor** (`cpu_triton_utils.py:55-56`); the kernel fills that temporary and the caller's buffer is never written. Every sibling wrapper in the file tracks `orig_dtype` and copies back — this one is the only one that doesn't.

Every production call hits it: `expand_batch_to_tokens` allocates `x.new_empty(num_tokens)` in x's dtype (`vllm/v1/sample/rejection_sampler.py:596`) and the three production inputs are float32 temperature, int32 top_k, float32 top_p (`vllm/v1/worker/gpu_input_batch.py:199-216`, call sites `rejection_sampler.py:537-560`). So on any CPU spec-decode run with a non-greedy request, `logits.div_(temperature)` (`:545`) and `apply_top_k_top_p` (`:565`) consume **uninitialized `new_empty` memory** — fresh zero pages give `div_(0)` → inf logits; arbitrary garbage gives the out-of-bounds crash previously observed downstream. A second layer: even with copy-back, `_ensure_int64(input_val)` truncates floats — temperature 0.7 → 0 → rewritten to 1 by the `replace_from=0` logic.

This path is live whenever Triton is absent (`cpu_model_runner.py:65-73` skips the patches only when `HAS_TRITON`; :115 installs this one) — i.e., pip installs without triton-cpu, macOS/aarch64.

### This was reported before, and the fix everyone believed in is the buggy code

PR #42638 reported exactly this corruption (uninitialized `expand_batch_to_tokens` output; third-party-confirmed on Xeon — OOB crash on Qwen's 151k vocab, **silently in-bounds on Llama's 128k**) via a different mechanism: the raw Triton kernel no-oping with no driver. A maintainer redirected the fix to the CPU `expand_kernel_impl` monkey-patch, and the PR was closed after #44991 (merged) fixed the Triton-CPU path, on the stated premise that the "existing C++ monkey-patch fallbacks" cover non-Triton environments. That premise was false: the expand fallback has never produced correct output for any of the three production dtypes.

### Reproduction (seconds; no model, no server, no GPU)

The script below, on a from-source CPU build at `605c3ddc` (`vllm 0.28.1rc1.dev315+g605c3ddcb.cpu`, torch 2.13.0, macOS arm64):

```
(a) int64 control      -> [10, 10, 20, 20, 20, 30]  expected [10, 10, 20, 20, 20, 30]  OK
(b) float32 out buffer -> [-777.0, -777.0, -777.0, -777.0, -777.0, -777.0]  UNTOUCHED (output discarded)
(c) fp32 temperature   -> [0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
    expected              [0.699999988079071, 0.699999988079071, 0.5, 0.5, 0.5, 0.8999999761581421]  DIVERGES (uninitialized new_empty)
(d) fp32 in, i64 out   -> [1, 1, 1, 1, 1, 1]  (0.7/0.5/0.9 all truncated to 0, then 'replaced' with 1)
(e) int32 top_k        -> [0, 0, 0, 0, 0, 0]
    expected              [5, 5, 40, 40, 40, 1]  DIVERGES
```

Arm (b) is deterministic proof: a sentinel-filled output buffer comes back untouched. Arm (d) isolates the truncation layer.

<details>
<summary>repro_expand_kernel_cpu.py</summary>

```python
# Repro: the CPU monkey-patch for `expand_kernel` discards its output and
# truncates float inputs — CPU spec decode with any non-greedy request
# consumes uninitialized memory as temperature / top_k / top_p.
#
# `vllm/utils/cpu_triton_utils.py::_expand_kernel_impl` passes
# `_ensure_int64(output)` to the C++ kernel. When the caller's buffer is not
# int64 (production passes float32 temperature/top_p and int32 top_k),
# `_ensure_int64` allocates a NEW tensor; the C++ kernel fills that temporary
# and the caller's `new_empty` buffer is returned uninitialized. Every sibling
# wrapper in the file copies back; this one doesn't. A second layer: float
# inputs are truncated to int64 (temperature 0.7 -> 0 -> replaced with 1).
#
# No model, no server, no GPU; seconds. Run on a CPU build of vLLM
# (path is live when Triton is absent, i.e. the C++ monkey-patch path).

import torch

import vllm  # noqa: F401
import vllm._C  # noqa: F401  (registers the torch.ops._C CPU kernels)
import vllm.utils.cpu_triton_utils as cpu_tl
import vllm.v1.sample.rejection_sampler as rs

# Monkey-patch exactly as vllm/v1/worker/cpu_model_runner.py:115 does when
# Triton is unavailable.
rs.expand_kernel = cpu_tl.expand_kernel

print(f"torch {torch.__version__}")
from importlib.metadata import version  # noqa: E402

print(f"vllm  {version('vllm')}")
print()

cu = torch.tensor([2, 5, 6], dtype=torch.int64)  # 3 requests -> 6 tokens

# --- Arm (a): int64 end-to-end (control: the C++ kernel itself is fine) ---
x = torch.tensor([10, 20, 30], dtype=torch.int64)
out = rs.expand_batch_to_tokens(x, cu, num_tokens=6)
print(f"(a) int64 control      -> {out.tolist()}  expected [10, 10, 20, 20, 20, 30]"
      f"  {'OK' if out.tolist() == [10, 10, 20, 20, 20, 30] else 'FAIL'}")

# --- Arm (b): output buffer is provably never written (deterministic) ---
# Call the shim directly with a float32 output buffer pre-filled with a
# sentinel. If the kernel wrote through, the sentinel would be gone.
out_f32 = torch.full((6,), -777.0, dtype=torch.float32)
cpu_tl._expand_kernel_impl(out_f32, torch.tensor([0.7, 0.5, 0.9]), cu, 0, 1)
untouched = bool((out_f32 == -777.0).all())
print(f"(b) float32 out buffer -> {out_f32.tolist()}  "
      f"{'UNTOUCHED (output discarded)' if untouched else 'written'}")

# --- Arm (c): the production temperature path (float32, new_empty) ---
# expand_batch_to_tokens allocates `x.new_empty(num_tokens)` in x's dtype;
# with the output discarded, callers receive uninitialized memory. This is
# exactly rejection_sampler.py:537-545: the result divides the logits.
temp = torch.tensor([0.7, 0.5, 0.9], dtype=torch.float32)
got = rs.expand_batch_to_tokens(temp, cu, num_tokens=6, replace_from=0, replace_to=1)
expected = torch.repeat_interleave(temp, torch.tensor([2, 3, 1]))
print(f"(c) fp32 temperature   -> {got.tolist()}")
print(f"    expected              {expected.tolist()}  "
      f"{'DIVERGES (uninitialized new_empty)' if not torch.equal(got, expected) else 'ok'}")

# --- Arm (d): float truncation, isolated from the discard bug ---
# With an int64 output buffer `_ensure_int64(output)` is the identity, so the
# kernel's writes ARE visible — and show the input-side truncation:
# temperature 0.7 -> int64 0 == replace_from -> replaced with 1.
out_i64 = torch.full((6,), -777, dtype=torch.int64)
cpu_tl._expand_kernel_impl(out_i64, temp, cu, 0, 1)
print(f"(d) fp32 in, i64 out   -> {out_i64.tolist()}  "
      f"(0.7/0.5/0.9 all truncated to 0, then 'replaced' with 1)")

# --- Arm (e): int32 top_k path (production dtype, gpu_input_batch.py:216) ---
topk = torch.tensor([5, 40, 1], dtype=torch.int32)
got_k = rs.expand_batch_to_tokens(topk, cu, num_tokens=6)
exp_k = torch.repeat_interleave(topk, torch.tensor([2, 3, 1]))
print(f"(e) int32 top_k        -> {got_k.tolist()}")
print(f"    expected              {exp_k.tolist()}  "
      f"{'DIVERGES' if not torch.equal(got_k, exp_k) else 'ok'}")
```

</details>

### What I cannot determine

- When the shim was introduced: `git log -S '_expand_kernel_impl'` bottoms out at my clone's graft boundary (depth 3,263), so introduction predates visible history.
- Whether the int64-only C++ kernel was a deliberate scope decision — the wrapper's discard + truncation is a bug either way.

### Fix

PR attached (posted alongside this issue): expand natively at the caller's dtype in the shim — the same `torch.repeat_interleave` semantics #42638 proposed, relocated to the monkey-patch layer where the maintainer asked for it. Regression test fails 3/4 arms before the fix, passes 4/4 after.

### Disclosure

This report was prepared with AI assistance (Claude Code); the analysis, measurements, and fix were reviewed and run by the submitter.

**Comment by m9ssah** [comment]
Is this unpatched? I'd like to take it up

**Comment by mosafariuk** [comment]
It's patched — the fix PR went up alongside this issue and is linked above: #55006 (regression test fails 3/4 arms on base, passes 4/4 after; already has an independent verification review). If you'd like to help, running the repro or the test on your platform and reporting results on the PR would be genuinely useful — more environments confirming the fix speeds up review.

**Comment by The-iyed** [comment]
Independent confirmation on a completed arm64 CPU build (`vllm 0.1.dev1+gbc162b3f9`, torch `2.13.0+cpu`, `HAS_TRITON = False`).

Rather than inferring from `new_empty` contents, I pre-filled the output buffer with a sentinel and called `_expand_kernel_impl` directly. If the kernel wrote through, the sentinel would be gone:

```
expand_kernel (output buffer dtype round-trip):
  PASS  expand out dtype=int64
  FAIL  expand out dtype=int32     got [-1,-1,-1,-1,-1,-1]  want [7,7,5,5,5,9]
  FAIL  expand out dtype=float32   got [-1,-1,-1,-1,-1,-1]  want [7,7,5,5,5,9]
```

The sentinel survives untouched for both non-int64 dtypes, so the caller's buffer is provably never written — exactly as this issue describes. int64 passes, consistent with `_ensure_int64` returning the same tensor in that case.

One thing worth adding: on current `main` this shim is not actually reached on arm64, because `_postprocess_triton` installs it by rebinding a module global that nothing reads since #56323 (details in #58214). So the corruption reported here is latent rather than live today — but it becomes live the moment that wiring is fixed. I hit exactly that in #58217: with the fallbacks correctly installed, the real call returns

```
expand_batch_to_tokens(temperature=[0.7, 0.5, 0.9]) over 6 tokens
  expected: [0.7, 0.7, 0.5, 0.5, 0.5, 0.9]
  got:      [0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
```

which is uninitialized memory used as temperature, so `logits.div_()` divides by zero.

So the two changes are complementary and #55006 is the one that has to land first — without it, fixing the wiring converts a loud `TypeError` into silent logit corruption.

AI assistance was used for this investigation.

**Comment by mosafariuk** [comment]
@The-iyed thank you — the sentinel-prefill is strictly cleaner than my `new_empty` inference, and I can confirm your matrix exactly on a macOS CPU build (torch 2.13.0, `HAS_TRITON=False`), including the `TypeError` arm.

Two additions from re-running it that sharpen the landing picture:

1. The dependency is mutual. `_launch_binding_plan` (jit_warmup_triton_helper.py:40-80) only binds the dispatch input to a kernel arg named `input`/`input_ptr`; #55006's shim keeps the historical `input_val` name, so on your corrected wiring it never binds — your rename is load-bearing for my fix too. #55006's body + your rename returns `[0.7, 0.7, 0.5, 0.5, 0.5, 0.9]`.

2. The diffs currently collide (`git merge-tree`: content conflict in `cpu_triton_utils.py`). Whichever lands second needs a small rebase — I'm happy to fold the rename into #55006 so #58217 rebases clean, or to rebase mine after yours lands; reviewer's pick.

One nuance on the zeros: they're uninitialized reads, so those values aren't guaranteed — which makes the failure mode worse, not better.

#55006 is mergeable against current `main` as of today and its 4-arm regression suite is green there.

**Comment by The-iyed** [comment]
@mosafariuk confirmed both, independently — and I think point 1 settles the landing order.

**The rename is load-bearing for #55006.** `kernel_arg()` builds its candidates from the *dispatch input* name only — `input`, `input_ptr`, and the uppercase pair — and matches them against the kernel's parameter names. So:

```
arg set {'input_val', 'output'}  ->  None      # never binds
arg set {'input',     'output'}  ->  'input'   # binds
```

And #55006's body with the rename applied gives exactly your matrix: `[0.7, 0.7, 0.5, 0.5, 0.5, 0.9]` from a `-777` sentinel prefill.

**The conflict reproduces, and it's smaller than it looks.** A three-way `git merge-file` against `main` gives exactly one conflict hunk — the *body* of `_expand_kernel_impl`. The signature rename merges cleanly:

```
 def _expand_kernel_impl(
     output,
     input,          <- merges clean, no conflict
     ...
 <<<<<<< ours
     torch.ops._C.expand_kernel_impl(_ensure_int64(output), _ensure_int64(input), ...
 =======
     counts = torch.diff(...)
     vals = torch.where(input_val == replace_from, ...
 >>>>>>> theirs
```

So folding it in is mechanical: take your body, swap the two `input_val` references to `input`.

**On the order — I'd take your first option, and not just for tidiness.** The two PRs are asymmetric:

- **#55006 alone is inert.** `cpu_tl.expand_kernel` is consumed at exactly one site, `cpu_model_runner.py:105`, and that assignment only reaches the dispatcher once #58217 lands. Until then the shim is unreachable, so a correct body is a safe no-op.
- **#58217 alone is not.** It makes the shim reachable while the body still discards its writes — the silent uninitialized reads feeding temperature/top_k/top_p, which as you say aren't guaranteed zeros.

So #55006 first, with the rename folded in, is the only order with no window where spec decode silently consumes garbage. Mine then rebases clean by dropping the rename hunk, which is the smaller rebase of the two anyway.

Happy to do it the other way if a reviewer prefers — but if #58217 goes first, it'd be worth landing them in the same batch rather than leaving the gap open.

(Unrelated: I've pushed a `ruff format` fix to #58217 — the repo's `pre-commit` is gated behind `pre-run-check` on my PR, so the formatting miss never surfaced in CI.)
